sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-generation.service.ts

```diff
@@ -0,0 +1,506 @@
+      // For each schedule calculate the total BC grants.
+      let bcTotalGrant = disbursementSchedule.disbursementValues.find(
+        (disbursementValue) =>
+          disbursementValue.valueType === DisbursementValueType.BCTotalGrant,
+      );
+      if (!bcTotalGrant) {
+        // If the 'BC Total Grant' is not present, add it.
+        bcTotalGrant = new DisbursementValue();
+        bcTotalGrant.creator = auditUser;
+        bcTotalGrant.valueCode = BC_TOTAL_GRANT_AWARD_CODE;
+        bcTotalGrant.valueType = DisbursementValueType.BCTotalGrant;
+        disbursementSchedule.disbursementValues.push(bcTotalGrant);
+      }
+      const bcTotalGrantValueAmount = disbursementSchedule.disbursementValues
+        // Filter all BC grants.
+        .filter(
+          (disbursementValue) =>
+            disbursementValue.valueType === DisbursementValueType.BCGrant,
+        )
+        // Sum all BC grants.
+        .reduce((previousValue, currentValue) => {
+          return previousValue + +currentValue.effectiveAmount;
+        }, 0)
+        .toString();
+      bcTotalGrant.valueAmount = bcTotalGrantValueAmount;
+      bcTotalGrant.effectiveAmount = bcTotalGrantValueAmount;
+    }
+  }
+
+  /**
+   * For a single student, check if there is an overaward balance, updates the awards
+   * with the deductions, if needed, and adjust the student overaward balance if a
+   * deduction happen.
+   * @param studentId student to be verified.
+   * @param studentSchedules student disbursements that are part of one e-Cert.
+   * @param studentsOverawardBalance overaward balance for the student.
+   * @param entityManager used to execute the commands in the same transaction.
+   */
+  private async applyOverawardsDeductionsForECertGeneration(
+    studentId: number,
+    studentSchedules: ECertDisbursementSchedule[],
+    studentOverawardBalance: AwardOverawardBalance,
+    entityManager: EntityManager,
+  ): Promise<void> {
+    // List of loan awards with the associated disbursement schedule.
+    // Filter possible awards that will not be disbursed due to a restriction.
+    // If the award will not be disbursed the overaward should not be deducted
+    // because the student will not be receiving any money for the award, in this
+    // case the BC Loan (BCSL).
+    const loanAwards = this.getAwardsByTypeWithRelatedSchedule(
+      studentSchedules,
+      LOAN_TYPES,
+    ).filter(
+      (loanAward) =>
+        !this.shouldStopFunding(
+          loanAward.relatedSchedule,
+          loanAward.awardValue,
+        ),
+    );
+    // Unique codes to allow the overawards deduction to happen in a sequential way.
+    const distinctValueCodes = [
+      ...new Set(loanAwards.map((loan) => loan.awardValue.valueCode)),
+    ];
+    for (const valueCode of distinctValueCodes) {
+      // Checks if the loan is present in multiple disbursements.
+      // Find all the values in all the schedules.
+      const loans = loanAwards.filter(
```

Comment:
mm... iam not getting this `distinctValueCodes` is the list of unique value codes in `loanAwards`. so, the value of `loans` will be always a list of `true` right?